### ISSUE TITLE
How to add a new datasource support which  has efcore integration

### ISSUE THREAD

**Issue by hatzhang** [ErikEJ/EFCorePowerTools#2361]
Suppose the support for the provider of Kingbase needed, which already has [ef core support](https://github.com/dotnetcore/EntityFrameworkCore.KingbaseES).
How should this begin, what is the estimated cost for doing this.

**Comment by ErikEJ** [comment]
There is no cost other than time - see this PR for an example of the relatively small amount of work involved: https://github.com/ErikEJ/EFCorePowerTools/pull/885/files (ignore the changes in the EFCorePowserTools folder)

**Comment by hatzhang** [comment]
Thanks very much.
The initial step seems quite simple, i checked in [a commit here](https://github.com/ErikEJ/EFCorePowerTools/compare/master...hatzhang:EFCorePowerTools:master). 
I ignored the changes in EFCorePowserTools folder, also I am have issues to compile the PowerToolsExtensionPack project, I had to remove it from the solution to build.

So, what is the next step, how to test or verify the behavior, and how to make it part of this great tool.

**Comment by ErikEJ** [comment]
@hatzhang Next step could be to create a PR with these initial changes. Does your provider only support EF Core 8?

After that, you can test via the EF Core 8 CLI tool, efcpt.8, by passing a connection string and provider name a command line parameters.

This file will need various updates to support the new provider: https://github.com/ErikEJ/EFCorePowerTools/blob/master/src/GUI/RevEng.Shared/Providers.cs

And the help instructions will need the new provider abbreviation added.

**Comment by hatzhang** [comment]
@ErikEJ  I created a PR. 
If I understood correctly, I should wait for the new nightly version of the `ErikEJ.EFCorePowerTools.Cli` which contains the change, then I can test via it? I tried build and run efcpt.8 at my local, it reports an error of `Unknown provider 'KingBase'`

The provider support both EF8 and EF6, but we are only going to use EF8.

I made changes to `Providers.cs` too, but I am not sure about the help instructions.

**Comment by ErikEJ** [comment]
You do not need to wait for a Nightly build  , you just need to make the CLI aware of the KingBase provider. Let me know if you need help with this

**Comment by ErikEJ** [comment]
You should be able to debug efcpt.8 by passing a connection string and provider name a command line parameters in the Debug options.

**Comment by hatzhang** [comment]
Yes, I can debug it, with those checked commits, by passing connection string and a right provider alias `kingbase`, the CLI is now aware the KingBase provider.
I copied an `efpt.config.json`, which we are using right now for reverse engineering against SQLServer, and for which it worked well.
But when running against KingBase, I am getting the following error.  I think this might be a provider issue? Not sure if there is a way to bypass it.

>System.Resources.MissingManifestResourceException: Could not find the resource
"Kdbndp.EntityFrameworkCore.KingbaseES.Properties.KdbndpStrings.resources" among the resources
"DotNetCore.EntityFrameworkCore.KingbaseES.Properties.KdbndpStrings.resources",
"DotNetCore.EntityFrameworkCore.KingbaseES.Strings.resources" embedded in the assembly
"DotNetCore.EntityFrameworkCore.KingbaseES", nor among the resources in any satellite assemblies for the specified
culture. Perhaps the resources were embedded with an incorrect name.
  at void System.Resources.ManifestBasedResourceGroveler.HandleResourceStreamMissing(string fileName)
  at ResourceSet System.Resources.ManifestBasedResourceGroveler.GrovelForResourceSet(CultureInfo culture,
     Dictionary<string, ResourceSet> localResourceSets, bool tryParents, bool createIfNotExists)
  at ResourceSet System.Resources.ResourceManager.InternalGetResourceSet(CultureInfo culture, bool createIfNotExists,
     bool tryParents)
  at string System.Resources.ResourceManager.GetString(string name, CultureInfo culture)
  at EventDefinitionBase Kdbndp.EntityFrameworkCore.KingbaseES.Internal.KdbndpResources.<>c.<LogFoundColumn>b__4_0(
     IDiagnosticsLogger logger)
  at TValue Microsoft.EntityFrameworkCore.Internal.NonCapturingLazyInitializer.EnsureInitialized<TParam,TValue>(ref
     TValue target, TParam param, Func<TParam, TValue> valueFactory)
  at FallbackEventDefinition Kdbndp.EntityFrameworkCore.KingbaseES.Internal.KdbndpResources.LogFoundColumn(
     IDiagnosticsLogger logger)
  at void Kdbndp.EntityFrameworkCore.KingbaseES.Internal.KdbndpLoggerExtensions.ColumnFound(
     IDiagnosticsLogger<Scaffolding> diagnostics, string tableName, string columnName, string dataTypeName, bool
     nullable, bool identity, string defaultValue, string computedValue)
  at void Kdbndp.EntityFrameworkCore.KingbaseES.Scaffolding.Internal.KdbndpDatabaseModelFactory.GetColumns(
     KdbndpConnection connection, IReadOnlyList<DatabaseTable> tables, string tableFilter, string internalSchemas,
     HashSet<string> enums, IDiagnosticsLogger<Scaffolding> logger)
  at IEnumerable<DatabaseTable> Kdbndp.EntityFrameworkCore.KingbaseES.Scaffolding.Internal.KdbndpDatabaseModelFactory.
     GetTables(KdbndpConnection connection, DatabaseModel databaseModel, Func<string, string, string> tableFilter,
     string internalSchemas, HashSet<string> enums, IDiagnosticsLogger<Scaffolding> logger)
  at DatabaseModel Kdbndp.EntityFrameworkCore.KingbaseES.Scaffolding.Internal.KdbndpDatabaseModelFactory.Create(
     DbConnection dbConnection, DatabaseModelFactoryOptions options)
  at DatabaseModel Kdbndp.EntityFrameworkCore.KingbaseES.Scaffolding.Internal.KdbndpDatabaseModelFactory.Create(string
     connectionString, DatabaseModelFactoryOptions options)
  at DatabaseModel RevEng.Core.TableListBuilder.GetDatabaseModel() in TableListBuilder.cs:131
  at List<TableModel> RevEng.Core.TableListBuilder.GetTableModels() in TableListBuilder.cs:43
  at void ErikEJ.EFCorePowerTools.Services.DisplayService.<>c__DisplayClass7_0`1.<Wait>b__0(StatusContext ctx) in
     DisplayService.cs:64
  at Task Spectre.Console.Status.<>c__DisplayClass14_0.<Start>b__0(StatusContext ctx) in Status.cs:44
  at void Spectre.Console.Status.<>c__DisplayClass16_0.<<StartAsync>b__0>d.MoveNext() in Status.cs:79
  at void Spectre.Console.Status.<>c__DisplayClass17_0`1.<<StartAsync>b__0>d.MoveNext() in Status.cs:120
  at void Spectre.Console.Progress.<>c__DisplayClass32_0`1.<<StartAsync>b__0>d.MoveNext() in Progress.cs:138
  at async Task<T> Spectre.Console.Internal.DefaultExclusivityMode.RunAsync<T>(Func<Task<T>> func) in
     DefaultExclusivityMode.cs:40
  at async Task<T> Spectre.Console.Progress.StartAsync<T>(Func<ProgressContext, Task<T>> action) in Progress.cs:121
  at async Task<T> Spectre.Console.Status.StartAsync<T>(string status, Func<StatusContext, Task<T>> func) in Status.cs:
     117
  at async Task Spectre.Console.Status.StartAsync(string status, Func<StatusContext, Task> action) in Status.cs:77
  at void Spectre.Console.Status.Start(string status, Action<StatusContext> action) in Status.cs:48
  at T ErikEJ.EFCorePowerTools.Services.DisplayService.Wait<T>(string message, Func<T> doFunc) in DisplayService.cs:59
  at List<TableModel> ErikEJ.EFCorePowerTools.HostedServices.ScaffoldHostedService.GetTablesAndViews() in
     ScaffoldHostedService.cs:187
  at async Task ErikEJ.EFCorePowerTools.HostedServices.ScaffoldHostedService.ExecuteAsync(CancellationToken
     stoppingToken) in ScaffoldHostedService.cs:43
  at async Task Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>b__15_1(IHostedService service, CancellationToken
     token)
  at async Task Microsoft.Extensions.Hosting.Internal.Host.ForeachService<T>(IEnumerable<T> services, CancellationToken
     token, bool concurrent, bool abortOnFirstException, List<Exception> exceptions, Func<T, CancellationToken, Task>
     operation)
  at void Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>g__LogAndRethrow|15_3(ref <>c__DisplayClass15_0 )
  at async Task Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
  at async Task<IHost> Microsoft.Extensions.Hosting.HostingAbstractionsHostBuilderExtensions.StartAsync(IHostBuilder
     hostBuilder, CancellationToken cancellationToken)
  at void ErikEJ.EFCorePowerTools.Program.<>c.<<MainAsync>b__1_1>d.MoveNext() in Program.cs:55
  at async Task<int> ErikEJ.EFCorePowerTools.Program.MainAsync(string[] args) in Program.cs:65
  at async Task<int> ErikEJ.EFCorePowerTools.Program.Main(string[] args) in Program.cs:22

**Comment by ErikEJ** [comment]
@hatzhang Yes, it is running now, this is a provider bug, cannot really assist with that.
